crates/ruff/src/importer/mod.rs

```diff
@@ -70,11 +70,16 @@ impl<'a> Importer<'a> {
         at: TextSize,
         semantic_model: &SemanticModel,
     ) -> Result<(Edit, String)> {
-        self.get_symbol(module, member, at, semantic_model)?
-            .map_or_else(
-                || self.import_symbol(module, member, at, semantic_model),
-                Ok,
-            )
+        match self.get_symbol(module, member, at, semantic_model) {
+            None => self.import_symbol(module, member, at, semantic_model),
+            Some(Resolution::Success(edit, binding)) => Ok((edit, binding)),
+            Some(Resolution::LateBinding) => {
+                bail!("Unable to use existing symbol due to late binding")
+            }
+            Some(Resolution::IncompatibleContext) => {
+                bail!("Unable to use existing symbol due to incompatible context")
+            }
+        }
```

Comment:
Nit: Would it make sense to define our own `Error` type here to allow callers to match on the reason or is this something that we don't need?

Reply:
I'm open to it -- do you have an example e.g. from Rome or elsewhere that I could use to learn how to implement that pattern?

Reply:
We have one in ruff :) 

https://github.com/charliermarsh/ruff/blob/4f9d944556910f834985b9fc75f968dc6fca2557/crates/ruff_formatter/src/diagnostics.rs#L1-L162

It's normally:

* Define an `enum`
* Implement `Error` and `Display`
* Maybe: Implement `From<OtherError>` if you want automatic error conversation when using the try operator (the try operator does `ReturnErrorType::from(error_from_the_tried_operation)`

Reply:
`thiserror` can also generate a bunch of the boilerplate in many cases

Reply:
I'm going to do this in a separate PR because I want to cover the `import_symbol` cases too.